src/db_options.rs

```diff
@@ -125,66 +134,105 @@ impl Env {
     pub fn set_low_priority_background_threads(&mut self, n: c_int) {
         unsafe {
-            ffi::rocksdb_env_set_low_priority_background_threads(self.inner, n);
+            ffi::rocksdb_env_set_low_priority_background_threads(self.0.inner, n);
         }
     }
 
     /// Sets the size of the bottom priority thread pool that can be used to
     /// prevent compactions from stalling memtable flushes.
     pub fn set_bottom_priority_background_threads(&mut self, n: c_int) {
         unsafe {
-            ffi::rocksdb_env_set_bottom_priority_background_threads(self.inner, n);
+            ffi::rocksdb_env_set_bottom_priority_background_threads(self.0.inner, n);
         }
     }
 
     /// Wait for all threads started by StartThread to terminate.
     pub fn join_all_threads(&mut self) {
         unsafe {
-            ffi::rocksdb_env_join_all_threads(self.inner);
+            ffi::rocksdb_env_join_all_threads(self.0.inner);
         }
     }
 
     /// Lowering IO priority for threads from the specified pool.
     pub fn lower_thread_pool_io_priority(&mut self) {
         unsafe {
-            ffi::rocksdb_env_lower_thread_pool_io_priority(self.inner);
+            ffi::rocksdb_env_lower_thread_pool_io_priority(self.0.inner);
         }
     }
 
     /// Lowering IO priority for high priority thread pool.
     pub fn lower_high_priority_thread_pool_io_priority(&mut self) {
         unsafe {
-            ffi::rocksdb_env_lower_high_priority_thread_pool_io_priority(self.inner);
+            ffi::rocksdb_env_lower_high_priority_thread_pool_io_priority(self.0.inner);
         }
     }
 
     /// Lowering CPU priority for threads from the specified pool.
     pub fn lower_thread_pool_cpu_priority(&mut self) {
         unsafe {
-            ffi::rocksdb_env_lower_thread_pool_cpu_priority(self.inner);
+            ffi::rocksdb_env_lower_thread_pool_cpu_priority(self.0.inner);
         }
     }
 
     /// Lowering CPU priority for high priority thread pool.
     pub fn lower_high_priority_thread_pool_cpu_priority(&mut self) {
         unsafe {
-            ffi::rocksdb_env_lower_high_priority_thread_pool_cpu_priority(self.inner);
+            ffi::rocksdb_env_lower_high_priority_thread_pool_cpu_priority(self.0.inner);
+        }
+    }
+
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct OptionsMustOutliveDB {
+    env: Option<Env>,
+    row_cache: Option<Cache>,
+    block_based: Option<BlockBasedOptionsMustOutliveDB>,
+}
+
+impl OptionsMustOutliveDB {
+    pub(crate) fn clone(&self) -> Self {
+        Self {
+            env: self.env.as_ref().map(Env::clone),
+            row_cache: self.row_cache.as_ref().map(Cache::clone),
+            block_based: self
+                .block_based
+                .as_ref()
+                .map(BlockBasedOptionsMustOutliveDB::clone),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct BlockBasedOptionsMustOutliveDB {
```

Comment:
Does it really need to be public rather than `pub(crate)`  ? If so then I'd rename it to `BlockBasedCache` or something because it seems weird name as for public API.

Reply:
It could be private!